src/main/asciidoc/reference/query-by-example.adoc

```diff
@@ -0,0 +1,167 @@
+
+Before getting started with Query by Example you need to have your entities set up.
+
+.Sample Person entity
+====
+[source,java]
+----
+public class Person {
+
+  @Id
+  private String id;
+  private String firstname;
+  private String lastname;
+  private Address address;
+
+  // … getters and setters omitted
+}
+----
+====
+
+We have a quite simple entity here that is mapped to the data store. You can use this entity to create an Example specification. By default, fields having `null` values are ignored, and strings are matched using the store specific defaults. Examples can be built by either using the `exampleOf` factory method or by using the <<query.by.example.builder,Example builder>>. Once the `Example` is constructed it becomes immutable.
+
+.The first Example specification
+====
+[source,xml]
+----
+Person person = new Person();                         <1>
+
+person.setFirstname("Dave");                          <2>
+
+Example<Person> example = Example.exampleOf(person);  <3>
+----
+<1> Create a new instance of the entity
+<2> Set the properties to query
+<3> Create an `Example`
+====
+
+
+NOTE: Property names of the sample object must correlate with the property names of the queried entity.
+
+.Query by Example using the MongoTemplate
+====
+[source,xml]
+----
+@Autowired
+MongoTemplate template;
+
+public List<Person> findPeople(Person sampleObject) {
+  return template.findByExample(Example.exampleOf(person));
+}
+----
+====
+
+The `findByExample` method accepts either the sample object or an `Example` object to query the data store. Spring Data uses the `Example` to create and execute a query.
+
+
+.Query by Example using the Repositories
+====
+[source, java]
+----
+public interface MongoRepository<Person, String> {
+
+  <S extends T> List<T> findAllByExample(Example<S> example);
+
+  <S extends T> List<T> findAllByExample(Example<S> example, Sort sort);
+
+  <S extends T> Page<T> findAllByExample(Example<S> example, Pageable pageable);
+
+  // … more functionality omitted.
+}
+----
+====
+
+[[query.by.example.builder]]
+== Example builder
+
+Examples are not limited to default settings. You can specify own defaults for string matching, null handling and property-specific settings using the example builder.
+
+.Query by Example builder
+====
+[source, java]
+----
+Example.newExampleOf(person)
+    .withStringMatcher(StringMatcher.ENDING)
+    .includeNullValues()
+    .withPropertySpecifier(
+            newPropertySpecifier("firstname").matchString(StringMatcher.CONTAINING).get())
+    .withPropertySpecifier(
+            newPropertySpecifier("lastname").matchStringsWithIgnoreCase().get())
+    .withPropertySpecifier(
+            newPropertySpecifier("address.city").matchStringStartingWith().get())
+    .get();
+----
+====
+
+Property specifier accepts property names and property paths separated by dots that are contained within the sample object. A `PropertySpecifier` allows setting string matching options, case-sensitivity,
```

Comment:
Property specifier accepts property names (e.g. "firstname" and "lastname"). You can navigate by chaining properties together with dots ("address.city"). You can tune it with matching options, case sensitivity...